### PR TITLE
Update debounce logic

### PR DESCRIPTION
**Pull request by devongovett**
This updates the debouncing logic to fix two issues:

* When changing only a single file, there was a minimum of 50ms of latency. We waited in case there were more future changes to batch. In order to optimize the common case of changing only a single file at a time, we now report the first event immediately, and batch subsequent events as before.
* As reported in #70, it was possible to never receive any events at all for a rapidly changing file (under the 50ms threshold) because we were always waiting for more. Now, we always emit at least one event every 500ms. Not sure if this threshold is too high, but I'd like to avoid triggering a Parcel build too often, for example while running a package manager.

### CODE REVIEW

**Comment by bpasero** [comment]
Looks like a consequence of this change is that an atomic write (= delete of a file and create of a file) is now indicated as a delete+create where previously it was just update. I think this is fine, will have to see in VS Code how to adjust, we have 1 failing test now when trying to update parcel.

**Comment by bpasero** [comment]
@devongovett I wonder though if this can further be improved, by combining both approaches:
* ensure that an event is emitted at least ever N milliseconds, fixing the issue that a constantly changing file is not emitting any event in `2.0.5`
* still do some event coalescing so that a rapidly deleted and recreated file (atomic write, probably git operations) result in an updated-event as in `2.0.5`

In other words:
* debounce as you did before for `50ms`
* still emit an event every `50ms` even when more events come in

**Comment by devongovett** [comment]
One of our goals for our last release of Parcel was to reduce the time from saving a file to building and seeing the result. Part of that was removing the artificial 50ms of latency added here for the common case, which is saving only one file at a time. Therefore, we now emit debounced updates on the leading rather than trailing edge.

There will always be cases where you will see a delete followed by a create, for example, if either of these operations take over 50ms. This change may make these more common, but hopefully it doesn't break your code. It also depends on the operating system, for example, macOS already does coalescing of events (your test above reports an update event on my system), and sometimes reports updates as creates, etc. We do our best to try to figure out what happened, but it's not always possible to be 100% accurate.

I suppose we could add back an option to debounce on the trailing edge if that would help you. What issues is this causing for vscode? Is it only a performance issue, or a correctness one?

**Comment by bpasero** [comment]
Hm yeah, I see the motivation and I think the new behaviour is fine for a file watcher library to produce the raw events as fast as possible so that upstream clients can react however they like. 

For VS Code it is a bit more complicate because we expose an API to our extensions to react to file events and part of that API is to declare whether you are interested in create, delete and/or updates. Most extensions only register for updates to trigger some action and never care about create/delete. If then a component triggers an update by doing delete+create (for whatever reasons), the extension would not function well anymore.

I can bring back debouncing into VS Code for parcel, we do the same for node.js already where there is no debouncing in node.js itself. The only ugly part is that I need to encode knowledge about parcels debouncing into our code, maybe the `50ms` should be a configuration option to break that.

Curious: if a file changes rapidly faster than the `50ms` debounce interval, will you emit events at least once in `50ms` or will the debounce logic result in no event until the changes are done?

**Comment by bpasero** [comment]
That said, I wonder if we should have a way to disable debouncing in parcel, given we do our own. Unless there is some logic in parcel that relies on debouncing to function well (e.g. on Linux, making sure to re-watch a deleted folder path when it comes back?), I think that would be great.

**Comment by devongovett** [comment]
> declare whether you are interested in create, delete and/or updates

Interesting. I'd be careful with relying on the event types too much as they can be inaccurate sometimes. If you have a way of tracking specific files you're interested in that's usually better. For example, in Parcel we track all files we've seen in our graph. If we get a "create" event for a file we already have, we know it's really an update. You could do something similar by tracking recently deleted files, and emitting updates to extensions when they are re-created. This may be more robust than time-based debouncing anyway if latency is unpredictable.

> if a file changes rapidly faster than the 50ms debounce interval, will you emit events at least once in 50ms or will the debounce logic result in no event until the changes are done?

The first event is emitted immediately, and subsequent events are batched. If there is a pause in events for more than 50ms, an event is emitted. Otherwise, events continue to be batched for a maximum of 500ms.

Anyway, I'll think about this more soon, and see about adding some options to allow consumers to control more of the debouncing behavior.
